## Section 1: Warm-up Questions


**Question 1.** 

Suppose you are asked to develop a frontend program for a database system that backs an ecommerce website. 
You need to write an `add_product` function, which allow people to add a new batch of products into the database. 

The information people must provide at the point of call includes the product name, the quantity, and the time when the batch is added. People can also possibly provide other custom specifics about the batch. 

Which of the following function signatures is the most appropriate for this desired function: 

- [x] `add_product(name, quantity, time, *args, **kwargs)`
- [ ] `add_product(name=None, quantity, time, **kwargs)`
- [ ] `add_product(name, quantity, time=None, *args)`
- [ ] `add_product(name, quantity, time, **kwargs)`




---

**Question 2. Element-wise text concatenation** 

Suppose you are given two lists, with their elements at the corresponding position representing the first name and the last name of the same person:

```python
first_names = ['Alice', 'Bob', 'Carol', 'David']
last_names = ['Alligator', 'Bear', 'Chimpanzee', 'Deer']
```

Write a lambda expression, name it `concat_names`, and call it with the two lists above to form a list of full names. The expected output is as follows:


```python
['Alice Alligator', 'Bob Bear', 'Carol Chimpanzee', 'David Deer']
```

In [42]:
first_names = ['Alice','Bob','Carol','David']
last_names = ['Alligator','Bear','Chimpanzee','Deer']


concat_names = list(map(lambda x, y: f"{x} {y}", first_names, last_names))
print(concat_names)




['Alice Alligator', 'Bob Bear', 'Carol Chimpanzee', 'David Deer']


---


**Question 3. Use of `lambda`s as function arguments** 

 

Suppose a music streaming service uses two lists to maintain a user's favorite songs and the corresponding ratings, respectively:

```python
songs = ['Back to Black', 'Poker Face', 'Yellow', 'Lolipop', 'All Too Well', 'Delicate', 'Moves Like Jagger']
ratings = [9.5, 8, 9, 8, 10, 7.5, 8]
```

1). Write code *without using loop statements* to create a playlist, which orders the user's favorite songs based on their ratings.  And if a tie is observed, include the lastest one first (the one closest to the end of the list).




The expected output should look like the following:

```python
['All Too Well', 'Back to Black', 'Yellow', 'Moves Like Jagger', 'Lolipop', 'Poker Face', 'Delicate']
```

Tips: use `enumerate()` and `sorted()` with a lambda expression.



In [43]:
# the solution we saw in the 2nd in-class exercise

songs = ['Back to Black', 'Poker Face', 'Yellow', 'Lolipop', 'All Too Well', 'Delicate', 'Moves Like Jagger']
ratings = [9.5, 8, 9, 8, 10, 7.5, 8]

rating_with_idx = [(rating, idx) for idx, rating in enumerate(ratings)]

[(songs[idx], rating) for rating, idx in sorted(rating_with_idx, reverse=True)]

[('All Too Well', 10),
 ('Back to Black', 9.5),
 ('Yellow', 9),
 ('Moves Like Jagger', 8),
 ('Lolipop', 8),
 ('Poker Face', 8),
 ('Delicate', 7.5)]

In [44]:
songs = ['Back to Black', 'Poker Face', 'Yellow', 'Lolipop', 
         'All Too Well', 'Delicate', 'Moves Like Jagger']

ratings = [9.5, 8, 9, 8, 10, 7.5, 8]

# Write one line of code below 
[songs[idx] for idx ,_ in sorted(enumerate(ratings), key=lambda x: (x[1],x[0]), reverse=True)]


['All Too Well',
 'Back to Black',
 'Yellow',
 'Moves Like Jagger',
 'Lolipop',
 'Poker Face',
 'Delicate']

2). What if we break the tie by sorting song titles lexicographically. The expected output should look like the following:

```python
[('All Too Well', 10),
 ('Back to Black', 9.5),
 ('Yellow', 9),
 ('Lolipop', 8),
 ('Moves Like Jagger', 8),
 ('Poker Face', 8),
 ('Delicate', 7.5)]
```

Tips: use `zip()` and `sorted()` with a lambda expression.


In [45]:
# Write your code below
sorted(zip(songs, ratings), key=lambda x: (-x[1],x[0]))



[('All Too Well', 10),
 ('Back to Black', 9.5),
 ('Yellow', 9),
 ('Lolipop', 8),
 ('Moves Like Jagger', 8),
 ('Poker Face', 8),
 ('Delicate', 7.5)]

---


**Question 4**


Assume that you are considering booking accommodation for a 1-week stay in Hong Kong on Airbnb, and you have multiple potential targets (e.g., dozens) under consideration.

Note that every accommodations on Airbnb is rated on a 5-star scale for each of the following 8 categories: Overall experience, cleanliness, accuracy, check-in, communication, location, value, and amenities.

Now, suppose that you manage to obtain the rating data for your target rentals, and maintain it in a Python list as follows:

```python
ratings = [[3, 4, 4, 5, 5, 3, 5, 3],
           [5, 5, 5, 3, 4, 4, 4, 5],
           [3, 5, 5, 2, 4, 4, 5, 2],
           [5, 5, 5, 5, 3, 2, 4, 5],
           [4, 5, 3, 3, 3, 4, 2, 4],
           [4, 5, 3, 3, 2, 3, 2, 5],
           [3, 4, 3, 4, 3, 4, 5, 4]]

```
where each sublist represents the ratings across 8 categories for one potential rental.

Write Python code to create a list that contains the indices of accommodations that receive the highest rating for each category.  

The expected output should look like the following:

```
[[2, 4], [2, 3, 4, 5, 6], [2, 3, 4], [1, 4], [1], [2, 3, 5, 7], [1, 3, 7], [2, 4, 6]]
```


Tips: 

1) Use `zip()` to reorganize the ratings by category;
2) Write a nested for loop to find the indices of the largest rating for each category.

In [46]:
ratings = [[3, 4, 4, 5, 5, 3, 5, 3],
           [5, 5, 5, 3, 4, 4, 4, 5],
           [3, 5, 5, 2, 4, 4, 5, 2],
           [5, 5, 5, 5, 3, 2, 4, 5],
           [4, 5, 3, 3, 3, 4, 2, 4],
           [4, 5, 3, 3, 2, 3, 2, 5],
           [3, 4, 3, 4, 3, 4, 5, 4]]

# iterate over ratings 
# for each column, select highest rating and check if any matches

# provide your code below
ls = []
for category in zip(*ratings):
    ls.append([i for i, rating in enumerate(category, start=1) if rating == max(category)])

print(ls)


[[2, 4], [2, 3, 4, 5, 6], [2, 3, 4], [1, 4], [1], [2, 3, 5, 7], [1, 3, 7], [2, 4, 6]]



---


## Section 2


For each question below, please write the corresponding code and then run the code to show the output. 







**Question 1.** 

Given a Python dictionary that maintians counts of dictinct words occurring in a document:

```python
word_count_dict = {'i': 2, 'like': 2, 'business': 3, 'analytics': 2, 'applications': 1, 
                   'development': 2, 'in': 3, 'python': 4, 'is': 3, 'interesting': 1, 
                   'a': 1, 'popular': 1, 'language': 1, 'application': 1, 'hard': 1}

```

Use `sorted()` with a `lambda` expression to produce the following output:

```python
[('python', 4), ('business', 3), ('in', 3), ('is', 3), ('analytics', 2), ('development', 2), ('i', 2), ('like', 2), ('a', 1), ('application', 1), ('applications', 1), ('hard', 1), ('interesting', 1), ('language', 1), ('popular', 1)]
```

Note that each dictionary was converted to a 2-element tuple, which represents the word count pair for a distinct word. All word-count pairs were sorted first by count in descending order and then by word in ascending order.

In [47]:
word_count_dict = {'i': 2, 'like': 2, 'business': 3, 'analytics': 2, 'applications': 1, 
                   'development': 2, 'in': 3, 'python': 4, 'is': 3, 'interesting': 1, 
                   'a': 1, 'popular': 1, 'language': 1, 'application': 1, 'hard': 1}

# Write your code below
sorted_word_list = sorted(word_count_dict.items(), key=lambda item: item[1], reverse=True)
print(sorted_word_list)





[('python', 4), ('business', 3), ('in', 3), ('is', 3), ('i', 2), ('like', 2), ('analytics', 2), ('development', 2), ('applications', 1), ('interesting', 1), ('a', 1), ('popular', 1), ('language', 1), ('application', 1), ('hard', 1)]



---



**Question 2**

Many programming languages implement a function called `sum()`, which returns the sum of an arbitrary sequence of numbers.

Define a function `mySum` that allows us to sum up any number of numeric values. For example:

- We should be able to call `mySum(1, 12, 89, 5)` with 4 arguments or `mySum(*range(100))` with 100 arguments.
- Calling `mySum()` with no arguments returns `0`.

The built-in function `sum()` is not allowed to use in the function body. Instead, use a for loop to tally the sum incrementally.

In [48]:
# Write your function definition below
def mySum(*nums):
    sum = 0
    if not nums:
        return 0
    else: 
        for num in nums:
            sum += num
        return sum

print(mySum(1, 12, 89, 5))
print(mySum(*range(100)))
        




107
4950



---

**Question 3**

Parallel processing of multiple sequences is a powerful functionality for data analysis.

Define a function `mySumParallel(seqs)` that applies `mySum()` defined in **Question 2** to an arbitrary collection of sequences of numeric values in parallel. This function returns a list of sums of all sequences in the passed-in collection.

For example, calling `mySumParallel(collection)` where `collection = [[1, 12, 89, 5], range(100), range(2, 9, 3)]` returns `[107, 4950, 15]`.



In [49]:
collection = [[1, 12, 89, 5], range(100), range(2, 9, 3)]

# Write your code below
def mySumParallel(seqs):
    new_seqs = list(map(lambda seq: mySum(*seq), seqs))
    return new_seqs
mySumParallel(collection)

    



[107, 4950, 15]

---

**Question 4**

Further define a function `myParallel(aggfunc, seqs)` that allows any aggregation function to be applied to an arbitrary collection of sequences of numeric values.

For example, calling `myParallel(mySum, collection)` yields `[107, 4950, 15]`, while calling `myParallel(max, collection)` yields `[89, 99, 8]`.

In [50]:
# Write your code below
def myParallel(aggfunc, seqs):
    if aggfunc == max:
        new_seqs = list(map(lambda seq: max(*seq), seqs))
        return new_seqs
    elif aggfunc == mySum:
        new_seqs = list(map(lambda seq: mySum(*seq), seqs))
        return new_seqs
    
print(myParallel(max, collection))
print(myParallel(mySum, collection))



[89, 99, 8]
[107, 4950, 15]


---

**Question 5**

Student A wants to select gifts for student B. Student A uses two dictionaries to record two types of ratings for options s/he considers: a dictionary that records the rating of how good a gift would be for student B (on a scale of 1-5), and a dictionary that records the rating of how much student B wants the gift (on a scale of 1-5). The rating of a gift that is not in either of the dictionaries is considered to be 0.

Student A wants to select the gifts with the maximal possible combined rating, where the combined rating is the sum of the rating of how good the gift would be for the student and the rating of how much the student wants the gift.

For example, the dictionaries can be:

```python
good_ratings = {"Python textbook": 5, "iPhone": 1, "iPad": 4, "AirPods": 4}
want_ratings = {"iPhone": 5, "A+ in ISOM3400": 5, "Python textbook": 3, "AirPods": 4, "Xbox": 5}
```

Here, the gifts student A wants to select are "Python textbook" and "AirPods", since the combined rating for them is 5+3=4+4=8, larger than any other one. The combined rating of "Xbox" is 5+0=5.


Write a function with the signature `select_gifts(good_ratings, want_ratings)` that returns a list
of all the gifts which have the highest combined rating of all the gifts, sorted in alphabetical order.


For example, for `good_ratings` and `want_ratings` as defined above,
`select_gifts(good_ratings, want_ratings)` should return `['AirPods', 'Python textbook']`.

In [51]:
# write your code below
good_ratings = {"Python textbook": 5, "iPhone": 1, "iPad": 4, "AirPods": 4}
want_ratings = {"iPhone": 5, "A+ in ISOM3400": 5, "Python textbook": 3, "AirPods": 4, "Xbox": 5}

# iterate over both dictionaries, create new dictionary to store unique items and their total score 
def select_gifts(good, want): 
    unique_dict = {item: rating for item, rating in good.items()}
    for item, rating in want.items():
        if item in unique_dict:
            unique_dict[item] += rating 
        else:
            unique_dict[item] = rating 
    highest_ls = [item for item, rating in unique_dict.items() if rating == max(unique_dict.values())]
    return sorted(highest_ls, key=lambda x: x[0])

select_gifts(good_ratings, want_ratings)
    






['AirPods', 'Python textbook']

---

**Question 6**


Suppose the organizer of the Hong Kong International Film Festival wants to determine the winner of a popularity award by conducting an online survey. In the survey form, respondents are asked to rate each nominated movie on a 5-point scale. The ratings collected by each form is stored as a flat Python list (e.g., `[4, 5, 3, 4, 5]` if there are 5 nominees), and the winner will be the movie that receives the highest average rating (or the largest sum of ratings).

Please define a custom Python function, `determine_winner(*forms)`, to perform needed calculations and return the index of the winner in the form (e.g., if the 2nd movie gets the highest average rating, return `2` as the output of call this function).

For example, calling it with the data collected by the following 4 forms:

```python
collected_forms = [[4, 5, 3, 4, 5], [3, 5, 2, 4, 4], [4, 5, 4, 4, 5], [3, 4, 4, 3, 3]]
determine_winner(*collected_forms)
```

produces `2`.


Assume that there will be no ties in the case of mass responses, and no missing values as well (so every form contains the same number of ratings).


In [53]:
# write your code below
# find sum of each movie and find highest, assuming no ties  
collected_forms = [[4, 5, 3, 4, 5], [3, 5, 2, 4, 4], [4, 5, 4, 4, 5], [3, 4, 4, 3, 3]]

def determine_winner(*forms):
    combined_ls = [sum(ls)/ len(ls) for ls in zip(*forms)]
    print(combined_ls)
    return combined_ls.index(max(combined_ls)) + 1

determine_winner(*collected_forms)


[3.5, 4.75, 3.25, 3.75, 4.25]


2